public/app/features/query/state/QueryRunner.ts

```diff
@@ -0,0 +1,154 @@
+import { getNextRequestId } from './PanelQueryRunner';
+import { preProcessPanelData, runRequest } from './runRequest';
+
+export class QueryRunner implements QueryRunnerSrv {
+  private subject: ReplaySubject<PanelData>;
+  private subscription?: Unsubscribable;
+  private lastResult?: PanelData;
+
+  constructor() {
+    this.subject = new ReplaySubject(1);
+  }
+
+  get(): Observable<PanelData> {
+    return this.subject.asObservable();
+  }
+
+  run(options: QueryRunnerOptions): void {
+    const {
+      queries,
+      timezone,
+      datasource,
+      panelId,
+      app,
+      dashboardId,
+      timeRange,
+      timeInfo,
+      cacheTimeout,
+      maxDataPoints,
+      scopedVars,
+      minInterval,
+    } = options;
+
+    console.log('subs', this.subscription);
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+
+    const request: DataQueryRequest = {
+      app: app ?? CoreApp.Unknown,
+      requestId: getNextRequestId(),
+      timezone,
+      panelId,
+      dashboardId,
+      range: timeRange,
+      timeInfo,
+      interval: '',
+      intervalMs: 0,
+      targets: cloneDeep(queries),
+      maxDataPoints: maxDataPoints,
+      scopedVars: scopedVars || {},
+      cacheTimeout,
+      startTime: Date.now(),
+    };
+
+    // Add deprecated property
+    (request as any).rangeRaw = timeRange.raw;
+
+    from(getDataSource(datasource, request.scopedVars))
+      .pipe(first())
+      .subscribe({
+        next: (ds) => {
+          // Attach the datasource name to each query
+          request.targets = request.targets.map((query) => {
+            if (!query.datasource) {
+              query.datasource = ds.name;
+            }
+            return query;
+          });
+
+          const lowerIntervalLimit = minInterval
+            ? getTemplateSrv().replace(minInterval, request.scopedVars)
+            : ds.interval;
+          const norm = rangeUtil.calculateInterval(timeRange, maxDataPoints, lowerIntervalLimit);
+
+          // make shallow copy of scoped vars,
+          // and add built in variables interval and interval_ms
+          request.scopedVars = Object.assign({}, request.scopedVars, {
+            __interval: { text: norm.interval, value: norm.interval },
+            __interval_ms: { text: norm.intervalMs.toString(), value: norm.intervalMs },
+          });
+
+          request.interval = norm.interval;
+          request.intervalMs = norm.intervalMs;
+
+          this.subscription = runRequest(ds, request).subscribe({
+            next: (data) => {
+              this.lastResult = preProcessPanelData(data, this.lastResult);
+              // Store preprocessed query results for applying overrides later on in the pipeline
+              this.subject.next(this.lastResult);
+            },
+            complete: () => {
+              if (this.subscription) {
+                this.subscription.unsubscribe();
+              }
+            },
```

Comment:
This should not be required, I do not think we do this anywhere else we use observables. Can we investigate if this is handled internally by observer.complete() ? 

I do not see this pattern in any rxjs example either